File: solutions/421/main.cpp
struct Trie {
    // left child = 0 
    Trie* left = nullptr; 
    // right child = 1 
    Trie* right = nullptr; 

    Trie() {}
}; 

class Solution {
private: 
    Trie* root = new Trie(); 
    static constexpr int HIGH_BIT = 30; 

public: 
    void add(int num) {
        Trie* cur = root; 
        for (int k = HIGH_BIT; k >= 0; k--) {
            int bit = (num >> k) & 1; 
            if (bit == 0) {
                if (!cur->left) {
                    cur->left = new Trie(); 
                }
                cur = cur->left; 
            } else {
                if (!cur->right) {
                    cur->right = new Trie(); 
                }
                cur = cur->right; 
            }
        }
    }

    int check(int num) {
        // check whether there is a number in the trie whose xor with num is maximum 
        Trie* cur = root; 
        int x = 0; 
        for (int k = HIGH_BIT; k >= 0; k--) {
            // check from highest bit to lowest bit 
            int bit = (num >> k) & 1; 
            if (bit == 0) {
                if (cur->right) {
                    cur = cur->right;
                    // the current bit can be 1, so add 1 
                    // eg checking first bit 
                    // ai = 0, aj = 1, ai xor aj = 1 
                    // x = 0 -> x = 1  
                    x = x * 2 + 1; 
                } else {
                    cur = cur->left; 
                    x *= 2; 
                }
            } else {
                if (cur->left) {
                    cur = cur->left; 
                    x = x * 2 + 1; 
                } else {
                    cur = cur->right; 
                    x *= 2; 
                }
            }
        }
        return x; 
    }

    int findMaximumXOR(vector<int>& nums) {
        int n = nums.size(); 
        int x = 0; 
        for (int i = 1; i < n; i++) {
            add(nums[i-1]); 
            x = max(x, check(nums[i]));
        }
        return x; 
    }
};

File: solutions/421/main.py
from typing import * 

# ref https://leetcode.cn/problems/maximum-xor-of-two-numbers-in-an-array/solutions/2511644/tu-jie-jian-ji-gao-xiao-yi-tu-miao-dong-1427d/?envType=daily-question&envId=2023-11-04

class Solution:
    def findMaximumXOR(self, nums: List[int]) -> int:
        ans = mask = 0 
        # bit length is the total length of bits 
        high_bit = max(nums).bit_length() - 1 
        
        for i in range(high_bit, -1, -1): 
            # bits lower than i are set to 0 
            mask |= 1 << i
            # check if current bit can be 1 
            target = ans | (1 << i)
            masked_nums = set([x & mask for x in nums])
            for x in masked_nums: 
                # a xor b = x 
                # then a xor b xor b = x xor b 
                # then a = x xor b
                # equivalent to find x + y = target  
                if target ^ x in masked_nums: 
                    ans = target  
                    break 
        
        return ans 
            
        